### PR TITLE
Action Validation framework and Explain API integration

### PR DESCRIPTION
**Pull request by jowg-amazon**
*Issue #, if available:*

#432 

*Description of changes:*

*CheckList:*
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/index-management/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: spi/src/main/kotlin/org.opensearch.indexmanagement.spi/indexstatemanagement/Validate.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    abstract fun executeValidation(indexName: String): Validate
+
+    abstract fun validatePolicy(): Boolean
+
+    abstract fun getUpdatedManagedIndexMetadata(currentMetadata: ManagedIndexMetaData, actionMetaData: ActionMetaData): ManagedIndexMetaData
+
+    enum class ValidationStatus(val status: String) : Writeable {
+        PASS("pass"),
+        REVALIDATE("revalidate"),
+        FAIL("fail");
```
Comment (bowenlan-amzn) [inline spi/src/main/kotlin/org.opensearch.indexmanagement.spi/indexstatemanagement/Validate.kt:35]: PASSED(passed)
RE_VALIDATING(re_validating)
FAILED(failed)

File: spi/src/main/kotlin/org.opensearch.indexmanagement.spi/indexstatemanagement/model/ManagedIndexMetaData.kt [hunk 2]
```diff
@@ -36,6 +36,7 @@ data class ManagedIndexMetaData(
     val stepMetaData: StepMetaData?,
     val policyRetryInfo: PolicyRetryInfoMetaData?,
     val info: Map<String, Any>?,
+    val validationMetaData: ValidationMetaData?,
```
Comment (bowenlan-amzn) [inline spi/src/main/kotlin/org.opensearch.indexmanagement.spi/indexstatemanagement/model/ManagedIndexMetaData.kt:39]: Just realised validation metadata/result can be treated as not different information but not part of managedindex metadata.

For explain API, we show the validation result directly in response

For runner, we also don't need to save it managedindex metadata, but can just use notification to send the validation result.

File: spi/src/main/kotlin/org.opensearch.indexmanagement.spi/indexstatemanagement/Validate.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+abstract class Validate(
+    val settings: Settings,
+    val clusterService: ClusterService
+) {
+
+    var validationStatus = ValidationStatus.PASS
+    var stepStatus = Step.StepStatus.STARTING
+    var validationMessage: String? = "Starting Validation"
+
+    abstract fun executeValidation(indexName: String): Validate
```
Comment (bowenlan-amzn) [inline spi/src/main/kotlin/org.opensearch.indexmanagement.spi/indexstatemanagement/Validate.kt:26]: nit: just execute is fine

File: spi/src/main/kotlin/org.opensearch.indexmanagement.spi/indexstatemanagement/Validate.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    abstract fun validatePolicy(): Boolean
```
Comment (bowenlan-amzn) [inline spi/src/main/kotlin/org.opensearch.indexmanagement.spi/indexstatemanagement/Validate.kt:28]: nit: executeAtPolicyCreation or give comment on how this function will be used

File: src/main/kotlin/org/opensearch/indexmanagement/IndexManagementPlugin.kt [hunk 5]
```diff
@@ -384,6 +386,7 @@ class IndexManagementPlugin : JobSchedulerExtension, NetworkPlugin, ActionPlugin
             .registerConsumers()
             .registerClusterConfigurationProvider(skipFlag)
         indexManagementIndices = IndexManagementIndices(settings, client.admin().indices(), clusterService)
+        validationService = ValidationService(settings, clusterService)
```
Comment (bowenlan-amzn) [inline src/main/kotlin/org/opensearch/indexmanagement/IndexManagementPlugin.kt:389]: I feel there's no need to init this here, we can just create this object in Explain or Runner.

**Review by bowenlan-amzn — APPROVED** [review]

**Review by Angie-Zhang — APPROVED** [review]

**Pull request merged**